lib/decompress/zstd_decompress_block.c

```diff
@@ -189,23 +217,53 @@ size_t ZSTD_decodeLiteralsBlock(ZSTD_DCtx* dctx,
                     break;
                 }
 
+                RETURN_ERROR_IF(litSize > 0 && dst == NULL, dstSize_tooSmall, "NULL not handled");
+                RETURN_ERROR_IF(expectedWriteSize < litSize, dstSize_tooSmall, "");
+                if (!streaming && dstCapacity > ZSTD_BLOCKSIZE_MAX + WILDCOPY_OVERLENGTH + litSize)
```

Comment:
I have the feeling I have already seen this logic before.
If yes, could it be factored ?

Reply:
The logic differs a bit between the three; in the huffman decoding case it is initially stored in one continuous block within dst (to avoid having to make complex changes to the huffman decoder), and then afterwards split partially into litExtraBuffer.  In the basic case, this is unnecessary so they literals are directly copied to the two split buffers, and in the rle case they are memset immediately in place.  I've added some comments to make it clearer.